File: Binary Tree Vertical Order Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    class Pair
    {
        TreeNode node;
        int dir;

        Pair(TreeNode node, int dir)
        {
            this.node = node;
            this.dir = dir;
        }
    }

    public List<List<Integer>> verticalOrder(TreeNode root) {
        
        if(root == null) return new ArrayList();

        HashMap<Integer, List<Integer>> map = new HashMap();
        List<List<Integer>> result = new ArrayList();

        Queue<Pair> queue = new LinkedList();
        queue.add(new Pair(root, 0));

        int maximum = 0, minimum = 0;

        while(!queue.isEmpty())
        {
            Pair p = queue.poll();

            if(!map.containsKey(p.dir)) map.put(p.dir, new ArrayList());

            map.get(p.dir).add(p.node.val);

            if(p.node.left != null)
            {
                queue.add(new Pair(p.node.left, p.dir - 1));
                minimum = Math.min(minimum, p.dir - 1);
            }

            if(p.node.right != null)
            {
                queue.add(new Pair(p.node.right, p.dir + 1));
                maximum = Math.max(p.dir + 1, maximum);
            }
        }

        for(int i = minimum; i <= maximum; i++) result.add(map.get(i));

        return result;
    }
}

File: BinaryTreeVerticalOrderTraversal.cpp
// Time Complexity - O(n), n is the number of nodes in the tree.
// Space Complexity - O(2n)[map + queue] ~ O(n), n is the number of nodes in the tree.
// Runs on Leetcode!

class Solution {
public:
    vector<vector<int>> verticalOrder(TreeNode* root) {
        vector<vector<int>> answer;
        if(!root)
            return answer;
        queue<pair<TreeNode*, int>> myQ;
        unordered_map<int, vector<int>> myMap;
        
        myQ.push({root, 0});
        int minC = 0; int maxC = 0;
        while(!myQ.empty()){
            pair currPair = myQ.front(); myQ.pop();
            myMap[currPair.second].push_back(currPair.first->val);
            
            if(currPair.first->left){
                myQ.push({currPair.first->left, currPair.second - 1});
                minC = min(minC, currPair.second - 1);
            }
            if(currPair.first->right){
                myQ.push({currPair.first->right, currPair.second + 1});
                maxC = max(maxC, currPair.second + 1);
            }
        }
        
        for(int i = minC; i <= maxC; i++)
            answer.push_back(myMap[i]);
        
        return answer;
    }
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */